src/cpp/session/modules/rmarkdown/SessionExecuteChunkOperation.hpp

```diff
@@ -0,0 +1,308 @@
+   
+private:
+   
+   enum OutputType { OUTPUT_STDOUT, OUTPUT_STDERR };
+   
+   void onStarted(ProcessOperations& operations)
+   {
+      isRunning_ = true;
+   }
+   
+   bool onContinue(ProcessOperations& operations)
+   {
+      return !terminationRequested_;
+   }
+   
+   void onExit(int exitStatus)
+   {
+      events().onChunkExecCompleted(docId_, chunkId_, notebookCtxId());
+      deregisterProcess();
+      isRunning_ = false;
+   }
+   
+   void onStdout(ProcessOperations& operations, const std::string& output)
+   {
+      onText(output, OUTPUT_STDOUT);
+   }
+   
+   void onStderr(ProcessOperations& operations, const std::string& output)
+   {
+      onText(output, OUTPUT_STDERR);
+   }
+   
+   void onText(const std::string& output, OutputType outputType)
+   {
+      using namespace core;
+      
+      // get path to cache file
+      FilePath target = chunkOutputFile(docId_, chunkId_, kChunkOutputText);
+      
+      // generate CSV to write
+      std::vector<std::string> values;
+      values.push_back(outputType == OUTPUT_STDOUT ? "1" : "2");
+      values.push_back(output);
+      std::string encoded = text::encodeCsvLine(values) + "\n";
+      
+      // write to cache
+      Error error = writeStringToFile(target, encoded);
+      if (error)
+         LOG_ERROR(error);
+      
+      // emit client event
+      enqueueChunkOutput(
+               docId_,
+               chunkId_,
+               notebookCtxId(),
+               kChunkOutputText,
+               target);
+   }
+   
+public:
+   
+   bool isRunning() const { return isRunning_; }
+   void terminate() { terminationRequested_ = true; }
+   bool terminationRequested() const { return terminationRequested_; }
+   const std::string& chunkId() const { return chunkId_; }
+   
+private:
+   bool isRunning_;
+   bool terminationRequested_;
+   std::string docId_;
+   std::string chunkId_;
+   ShellCommand command_;
+   
+private:
+   
+   typedef std::map<
+      std::string,
+      boost::shared_ptr<ExecuteChunkOperation>
+   > ProcessRegistry;
+   
+   static ProcessRegistry& registry()
+   {
+      static ProcessRegistry instance;
+      return instance;
+   }
+   
+   void registerProcess()
+   {
+      registry()[docId_ + "-" + chunkId_] = shared_from_this();
+   }
+   
+   void deregisterProcess()
+   {
+      registry().erase(docId_ + "-" + chunkId_);
+   }
+   
+public:
+   static boost::shared_ptr<ExecuteChunkOperation> getProcess(
+         const std::string& docId,
+         const std::string& chunkId)
+   {
+      return registry()[docId + "-" + chunkId];
+   }
+};
+
+core::Error runChunk(const std::string& docId,
+                     const std::string& chunkId,
+                     const std::string& engine,
+                     const std::string& code)
+{
+   using namespace core;
+   typedef core::shell_utils::ShellCommand ShellCommand;
+   
+   // write code to temporary file
+   FilePath scriptPath = module_context::tempFile("chunk-code", "");
```

Comment:
Since someone might run a chunk many times in a single session, it seems like a lot of these temporary files could accumulate -- it is probably a bit better to clean up these files ourselves when the process exits (I'll make that change)